src/torchrunx/logging_utils.py

```diff
@@ -27,7 +28,44 @@ def record_factory(*args, **kwargs):
     logger.addHandler(SocketHandler(host=logger_hostname, port=logger_port))
 
 
-class LogRecordSocketReceiver(TCPServer):
+def default_handlers(
+    hostnames: list[str], workers_per_host: list[int], log_dir: str = "./logs"
+) -> list[Handler]:
+    handlers = []
+
+    timestamp = datetime.datetime.now().isoformat(timespec="seconds")
+
+    def make_handler(hostname: str, rank: int | None = None, stream: bool = False) -> Handler:
+        if stream:
+            handler = logging.StreamHandler()
+        else:
+            handler = logging.FileHandler(
+                f"{log_dir}/{timestamp}-{hostname}{'' if rank is None else f'[{rank}]'}.log"
+            )
+
+        def handler_filter(record: logging.LogRecord) -> bool:
+            return record.hostname == hostname and record.worker_rank == rank
+
+        handler.addFilter(handler_filter)
+        handler.setLevel(logging.DEBUG)
+        formatter = logging.Formatter(
+            "%(asctime)s:%(levelname)s:%(hostname)s:worker-%(worker_rank)s:%(message)s"
+        )
+        handler.setFormatter(formatter)
+
+        return handler
+
+    for i, hostname in enumerate(hostnames):
+        handlers.append(make_handler(hostname=hostname))
+        for j in range(workers_per_host[i]):
+            handlers.append(make_handler(hostname=hostname, rank=j))
+            if i == 0 and j == 0:
+                handlers.append(make_handler(hostname=hostname, rank=j, stream=True))
+
+    return handlers
+
+
+class LogRecordSocketReceiver(ThreadingTCPServer):
```

Comment:
Would you kindly explain to me why we might prefer `ThreadingTCPServer` over `TCPServer` -- if it matters? I think handling should be pretty fast, so I would prefer to process the log records synchronously (rather than ending up with race-conditions, etc).

Reply:
I changed it -- please let me know if you have objections.

Reply:
Okay, my understanding is that, the Launcher might call `log_process.kill()` before the `TCPServer` is finished handling the current log records. I think `ThreadingTCPServer` is your solution to making sure the log records are handled more quickly?

I am not sure if this resolves the issue completely (but significantly reduces the likelihood). I am wondering if we can keep `TCPServer` (instead of `ThreadingTCPServer`) and just call `log_receiver.shutdown()` to block until the remaining records are handled before `log_process.kill()`. But in my testing so far, the process seems to hang on the `shutdown()` call.

Reply:
For the record:

- We found that `ThreadingTCPServer` appears to handle log requests in the correct (chronological) order (more often than `TCPServer`).
- We overrode `shutdown()` and included a few second timeout, because it was stalled (still not sure why). Not sure whether this is a graceful shutdown. But it appears to be enough to "flush" the logs.